File: 0015-3sum/0015-3sum.go
import "slices"

func threeSum(nums []int) [][]int {
  n := len(nums)-1
  ans := [][]int{}
  slices.Sort(nums)
  for i := range n-1 {
    if i > 0 && nums[i] == nums[i-1] { continue }
    target, j, k := -nums[i], i+1, n
    for j < k {
      switch {
        case nums[j] + nums[k] == target:
          ans = append(ans, []int{nums[i], nums[j], nums[k]})
          j++
          for j<k && nums[j] == nums[j-1] { j++ }
        case nums[j] + nums[k] < target:
          j++
        default:
          k--
      }
    }
  }
  return ans
}

File: 0015-3sum/0015-3sum.py
class Solution:
  def threeSum(self, nums: List[int]) -> List[List[int]]:
    r = len(nums)-1
    ans = []
    nums.sort()
    for l in range(r-1):
      if not l or nums[l] != nums[l-1]:
        target, j, k = -nums[l], l+1, r
        while j < k:
          if nums[j] + nums[k] == target:
            if nums[j] != nums[j]-1 and (k == r or nums[k] != nums[k+1]):
              ans.append([nums[l], nums[j], nums[k]])
          if nums[j] + nums[k] < target:
            j += 1
          else:
            k -= 1
    return ans
